Implement header interface for block and add block exchange service

Manav-Aggarwal opened the pull request:
<!--
Please read and fill out this form before submitting your PR.

Please make sure you have reviewed our contributors guide before submitting your
first PR.
-->

## Overview

Resolves #1005. resolves #1021 

<!-- 
Please provide an explanation of the PR, including the appropriate context,
background, goal, and rationale. If there is an issue with this information,
please provide a tl;dr and link the issue. 
-->

## Checklist

<!-- 
Please complete the checklist to ensure that the PR is ready to be reviewed.

IMPORTANT:
PRs should be left in Draft until the below checklist is completed.
-->

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+	"github.com/rollkit/rollkit/types"
+)
+
+type BlockExchangeService struct {
```
We should get in the habit of better documentation of our structs. We should explain what the struct is responsible for so that developer have the right context to be able to understand why it has the fields it does. 

This then allows us to add more helpful comments on the fields to explain what they are responsible for. 

Especially with a project like rollkit that is meant to support so many different DAs, EEs, rollup types, etc, I think over documenting our code is going to help us in the long run in terms of catching edge cases and understanding the impact of design changes.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
```
why is this not part of `Start`?

Manav-Aggarwal replied:
it is called by `Start`

MSevey replied:
resolving this, deferring to my other comment.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
```
nit: I think always trying to keep code more inline is more readable.

```suggestion
	if trustedBlock == nil {
                return
	}
	if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
		bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
	}
```

Additionally, why do we not return the error and let the caller decide how to handle it? 
I think if we are handling errors within a function and not returning anything, we should have a clear docstring explaining the logic of the function and how errors are handled. 
Or we should defer the logic to the caller and just return an error.

Manav-Aggarwal replied:
Talked about this with Ganesh, this method was used as a goroutine before which is why he didn't return errors and only logged them when writing HeaderExchange. We decided that this method can be blocking so modified it to return errors as well.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
```
For all these, if there is an error we should return. If we can't init nor start, we can't do anything else, so all those errors would just be spamming the logs.

Manav-Aggarwal replied:
Made it to return the error in case of a failed init or start but we still log errors in case it failed to broadcast. Since this is used in `blockPublishLoop` which just relays blocks, it's fine to log errors then.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
```
nit: if you are define the error immediately, just do it in the return
```suggestion
func (bExService *BlockExchangeService) Start() (err error) {
```

MSevey replied:
but also, looking at the rest of the method, I don't think you need to define it at all.  
So the line could just be deleted

```suggestion
func (bExService *BlockExchangeService) Start() error {
```

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
```
We should define variables where they are needed, so this should be moved to line 121. 

On line 110 just change `=` to `:=`

Manav-Aggarwal replied:
Modified.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+
+	_, _, network := bExService.p2p.Info()
+	if bExService.p2pServer, err = newP2PServer(bExService.p2p.Host(), bExService.blockStore, network); err != nil {
+		return err
```
might as well add some context to this error since you do it for every other one :-)

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+	if bExService.ex, err = newP2PExchange(bExService.p2p.Host(), peerIDs, network, bExService.genesis.ChainID, bExService.p2p.ConnectionGater()); err != nil {
+		return err
+	}
+	if err = bExService.ex.Start(bExService.ctx); err != nil {
```
even though `err` has been defined at this point, I think it is a better practice to keep these inline variables scoped to avoid errors if things are refactored. 
```suggestion
	if err := bExService.ex.Start(bExService.ctx); err != nil {
```

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	// Check if the headerstore is not initialized and try initializing
+	if bExService.blockStore.Height() > 0 {
```
[non blocking] 
In general I think it is good practice to make use of `isInitialized` type methods here, so that how we are checking the state of initialization is controlled and can be unit tested. Additionally it reduces the surface area for bugs if different methods are checking a state differently.

Manav-Aggarwal replied:
added an `isInitialized `method

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		bExService.syncerStatus.started = true
```
this is a perfect example of why the status should be set in the `Start` method. Here we are editing the field outside of a mutex lock.

Manav-Aggarwal replied:
Talked about this with Ganesh, and he says that this field is never touched by another method while we do this. I think it's better practice to just make this a method.

MSevey replied:
> he says that this field is never touched by another method while we do this. 

I don't think this should be how we think about it. This may be true today, but this could change in the future. We should write code that is future proof, which means protecting against potential future data races. If a struct has a mutex, there is no reason to not use imo. There is virtually zero performance impact and making this a habit will lead to more robust code.

Manav-Aggarwal replied:
Agreed, which is why I made a method that uses the mutex :) cc @gupadhyaya for the above comment

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+			}
+		}
+	}
+	go bExService.tryInitBlockStoreAndStartSyncer(bExService.ctx, trustedBlock)
```
what is the rationale behind not blocking on this?

Manav-Aggarwal replied:
Talked to Ganesh about it and made this blocking.

Manav-Aggarwal commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		bExService.syncerStatus.started = true
+		return nil
```
added a comment. "can return here since syncer is initialized"

Manav-Aggarwal commented on node/block_exchange.go:
```diff
@@ -0,0 +1,226 @@
+	ctx    context.Context
+}
+
+func NewBlockExchangeService(ctx context.Context, store ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, logger log.Logger) (*HeaderExchangeService, error) {
```
added checks for `genesis` or `p2p` being `nil`

gupadhyaya commented on block/manager.go:
```diff
@@ -60,6 +60,7 @@ type Manager struct {
 	daHeight uint64
 
 	HeaderCh chan *types.SignedHeader
+	BlockCh  chan *types.Block
```
you need to initialize the channel?

gupadhyaya replied:
also need to do `m. BlockCh <- &block` in `publishBlock`?

Manav-Aggarwal replied:
Added

MSevey commented on node/exchange.go:
```diff
@@ -0,0 +1,24 @@
+	syncerStatus.started = true
+}
+
+func (syncerStatus *SyncerStatus) getStarted() bool {
```
[non-blocking] 
nit: typically `get` is associated with some arbitrary value while `is` is more often associated with a boolean. So `isStarted` would make more sense in how we are using this method.

MSevey commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+	genesis      *tmtypes.GenesisDoc
+	p2p          *p2p.Client
+	ex           *goheaderp2p.Exchange[*types.Block]
+	syncer       *goheadersync.Syncer[*types.Block]
+	sub          *goheaderp2p.Subscriber[*types.Block]
+	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
```
[non-blocking]
I think we can improve how we are handling the syncer status. First, as a nit, I would group the related fields to make it more explicit.
```suggestion
	sub          *goheaderp2p.Subscriber[*types.Block]
	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
	blockStore   *goheaderstore.Store[*types.Block]
	
	syncer       *goheadersync.Syncer[*types.Block]
	syncerStatus *SyncerStatus
```

Second, I think it makes since to have a helper method of the `BlockExchangeService` called `StartSyncer`:

```go
func (bes *BlockExchangeService) StartSyncer() error {
  bes.syncerStatus.mu.Lock()
  defer bes.syncerStatus.mu.Unlock()
  if bes.syncerStatus.started {
    return nil
  }
  err := bes.syncer.Start()
  if err != nil {
    return err
  }
  bes.syncerStatus.started = true
  return nil
}
```

MSevey reviewed (changes requested):
Looks good. Would look even better with tests 😉

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("failed to start syncer after initializing block store")
+		}
```
do we need to set the syncer status as started here?

Manav-Aggarwal replied:
Used the method as Matt mentioned, so that sets the syncer as well

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+func (bExService *BlockExchangeService) Start() error {
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.Block](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID+"-block")
```
can use variables for: `bExService.genesis.ChainID+"-block"` and `network+"-block"`

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+			if trustedBlock, err = bExService.ex.GetByHeight(bExService.ctx, uint64(bExService.genesis.InitialHeight)); err != nil {
+				// Full/light nodes have to wait for aggregator to publish the genesis block
+				// proposing aggregator can init the store and start the syncer when the first block is published
+				bExService.logger.Info("failed to fetch the genesis block", "error", err)
```
lets return error here, which will prevent the start of secondary aggregator and full nodes.

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+				// proposing aggregator can init the store and start the syncer when the first block is published
+				bExService.logger.Info("failed to fetch the genesis block", "error", err)
+			}
+		}
```
let's move this inside the outer if: `return bExService.initBlockStoreAndStartSyncer(bExService.ctx, trustedBlock)`

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+			}
+		}
+	}
+	if trustedBlock == nil {
```
remove `if` and simply return `nil`

gupadhyaya commented on node/exchange.go:
```diff
@@ -0,0 +1,24 @@
+}
+
+func (syncerStatus *SyncerStatus) isStarted() bool {
+	syncerStatus.m.Lock()
```
use RLock as it is read-only.

gupadhyaya commented on node/full.go:
```diff
@@ -239,7 +246,26 @@ func (n *FullNode) headerPublishLoop(ctx context.Context) {
-			n.hExService.writeToHeaderStoreAndBroadcast(ctx, signedHeader)
+			err := n.hExService.writeToHeaderStoreAndBroadcast(ctx, signedHeader)
+			if err != nil {
+				// failed to init or start headerstore
```
you should log the err here as return has no effect because `headerPublishLoop` is run in its own goroutine

gupadhyaya commented on node/full.go:
```diff
@@ -239,7 +246,26 @@ func (n *FullNode) headerPublishLoop(ctx context.Context) {
+		case block := <-n.blockManager.BlockCh:
+			err := n.bExService.writeToBlockStoreAndBroadcast(ctx, block)
+			if err != nil {
+				// failed to init or start blockstore
```
same as headerPublishLoop, log the error.

gupadhyaya commented on node/header_exchange.go:
```diff
@@ -175,9 +178,10 @@ func (hExService *HeaderExchangeService) Start() error {
+	if trustedHeader == nil {
+		return nil
+	}
+	return hExService.initHeaderStoreAndStartSyncer(hExService.ctx, trustedHeader)
```
apply the same comments as in block exchange service. move `initHeaderStoreAndStartSyncer ` inside the outer-if.

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+}
+
+func (b *Block) ChainID() string {
+	return b.SignedHeader.BaseHeader.ChainID
```
just call `b.SignedHeader.ChainID()` instead of field access.

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+}
+
+func (b *Block) Height() int64 {
+	return int64(b.SignedHeader.BaseHeader.Height)
```
`b.SignedHeader. Height()`

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+}
+
+func (b *Block) Time() time.Time {
+	return time.Unix(int64(b.SignedHeader.BaseHeader.Time), 0)
```
`b.SignedHeader.Time()`

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+
+func (b *Block) Verify(untrst header.Header) error {
+	//TODO: Update with new header verify method
+	untrstH, ok := untrst.(*Header)
```
the type assertion should be for `*Block` not `*Header`.

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+		panic(fmt.Errorf("%T is not of type %T", untrst, &b.SignedHeader.Header))
+	}
+	// sanity check fields
+	if err := verifyNewHeaderAndVals(&b.SignedHeader.Header, untrstH); err != nil {
```
`untrstH.SignedHeader.Header`

Manav-Aggarwal commented on node/block_exchange.go:
```diff
@@ -0,0 +1,233 @@
+
+	_, _, network := bExService.p2p.Info()
+	var err error
+	if bExService.p2pServer, err = newBlockP2PServer(bExService.p2p.Host(), bExService.blockStore, network+"-block"); err != nil {
```
For sure, created an issue earlier for it here: https://github.com/rollkit/rollkit/issues/1020

gupadhyaya replied:
you should create an issue in go-header repo and link it here, as the PR to fix will be made in go-header.

Manav-Aggarwal replied:
Created relevant issue in go-header: https://github.com/celestiaorg/go-header/issues/54

gupadhyaya commented on types/block.go:
```diff
@@ -61,3 +64,46 @@ type Signature []byte
+		panic(fmt.Errorf("%T is not of type %T", untrst, &b))
+	}
+	// sanity check fields
+	if err := untrstB.ValidateBasic(); err != nil {
```
you should use `VerifyNewHeaderAndVals(b.SignedHeader.Header, untrstB.SignedHeader.Header)`

gupadhyaya commented on node/block_exchange.go:
```diff
@@ -0,0 +1,246 @@
+func (bExService *BlockExchangeService) Start() error {
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	ChainIDBlock := bExService.genesis.ChainID + "-block"
```
`chainIDBlock`

gupadhyaya commented:
@Manav-Aggarwal do you plan to add some basic tests for making sure the exchange is functional?

Manav-Aggarwal commented:
@gupadhyaya currently blocked on https://github.com/rollkit/rollkit/pull/992 for adding tests. Happy to add them in this PR or in a future PR.

MSevey commented on go.mod:
```diff
@@ -187,6 +187,7 @@ require (
 )
 
 replace (
+	github.com/celestiaorg/go-header => github.com/gupadhyaya/go-header v0.0.0-20230623130312-b9d75475a548
```
?? was this for when a PR was opened?

gupadhyaya replied:
nope. last night fix commit added that. it will be updated once https://github.com/celestiaorg/go-header/pull/57 is merged.

Manav-Aggarwal replied:
Yes, this change was part of the fix from @gupadhyaya and for which this PR: https://github.com/celestiaorg/go-header/pull/57 was opened. Ideally, we have a new go-header version once https://github.com/celestiaorg/go-header/pull/57 is merged and we can use that new go-header version for this rollkit PR.

nashqueue replied:
https://github.com/celestiaorg/go-header/pull/57 is merged we can do a release in go-header

Manav-Aggarwal commented:
Extracted the changes made to refactor `header_exchange.go` into https://github.com/rollkit/rollkit/pull/1028 since they make a test, `LazyAggregator`, in `full_node_integration_test.go` fail. Also, refactored tests to reuse the `TestHeaderExchange` tests in a new test called `TestBlockExchange` for testing Block Exchange Service.

MSevey commented on node/full_node_integration_test.go:
```diff
@@ -172,33 +172,39 @@ func TestLazyAggregator(t *testing.T) {
+	testSingleAggregatorSingleFullNode(t, false)
+	testSingleAggregatorTwoFullNode(t, false)
+	testSingleAggregatorSingleFullNodeTrustedHash(t, false)
+	testSingleAggregatorSingleFullNodeSingleLightNode(t)
```
What makes this test different that we don't use the `useBlockExchange` bool?

gupadhyaya replied:
i think the light node does not have block exchange service, it only has header exchange.
